process: add process.getBuiltinModule(id)

joyeecheung opened the pull request:
`process.getBuiltinModule(id)` provides a way to load built-in modules
in a globally available function. ES Modules that need to support
other environments can use it to conditionally load a Node.js built-in
when it is run in Node.js, without having to deal with the resolution
error that can be thrown by `import` in a non-Node.js environment or
having to use dynamic `import()` which either turns the module into
an asynchronous module, or turns a synchronous API into an asynchronous one.

```mjs
if (globalThis.process?.getBuiltinModule) {
  // Run in Node.js, use the Node.js fs module.
  const fs = globalThis.process.getBuiltinModule('fs');
  // If `require()` is needed to load user-modules, use createRequire()
  const module = globalThis.process.getBuiltinModule('module');
  const require = module.createRequire(import.meta.url);
  const foo = require('foo');
}
```

If `id` specifies a built-in module available in the current Node.js process,
`process.getBuiltinModule(id)` method returns the corresponding built-in
module. If `id` does not correspond to any built-in module, `undefined`
is returned.

`process.getBuiltinModule(id)` accepts built-in module IDs that are recognized
by `module.isBuiltin(id)`.

The references returned by `process.getBuiltinModule(id)` always point to
the built-in module corresponding to `id` even if users modify
`require.cache` so that `require(id)` returns something else.


Fixes: https://github.com/nodejs/node/issues/52599

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

joyeecheung commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+added: REPLACEME
+-->
+
+* `id` {string} ID of the built-in module being requested. Must not contain the `node:` prefix.
```
I am not actually sure if forbidding `node:` is a good idea, but then it seems a bit weird to allow it since the whole point of the `node:` prefix is to prevent ambiguity in `require()` that may be used to load user modules. But with `process.getBuiltin()` which can only be used to load built-ins, such ambiguity does not exist...

GeoffreyBooth replied:
I don’t see the point on erroring for the `node:` prefix. Permitting it allows using the same set of builtin specifiers for this and for `import()`.

joyeecheung replied:
The weird exception arises when you can `process.getBuiltin('test')` but not `require('test')` (or they don't return the same thing if you don't happen to have a module called `test`).

joyeecheung replied:
Another way to make it less weird would be to only allow `node:` IDs, but then it somehow still feels weird - this is only for loading built-ins, why all the `node:` to disambiguate?

GeoffreyBooth replied:
Per https://nodejs.org/api/module.html#moduleisbuiltinmodulename:

```js
import { isBuiltin } from 'node:module';
isBuiltin('node:fs'); // true
```

If this works, then `process.getBuiltin('node:fs')` needs to work too.

joyeecheung replied:
That goes back to the `node:test` exception, because `module.isBuiltin('test')` returns false. That means `process.getBuiltin('test')` should throw?

joyeecheung replied:
And if `process.getBuiltin('test')` throws, that encourage people again to do `node:` always even with `process.getBuiltin()`, which then feels a bit strange because `process.getBuiltin()` does not load any user modules and there is nothing to disambiguate...

GeoffreyBooth replied:
That's fine. It's better to be consistent with the other builtin APIs.

joyeecheung replied:
> It's better to be consistent with the other builtin APIs.

There is the more widely used `module.builtinModules` (also used in the tests) which doesn't include `node:` names...a quick search with `module.builtinModules.includes` that many popular projects are using it to determine whether an ID is a builtin https://github.com/search?q=module.builtinModules.includes&type=code

GeoffreyBooth replied:
> There is the more widely used `module.builtinModules` (also used in the tests) which doesn’t include `node:` names

That includes neither `test` nor `node:test`. But I don’t think we want to have `getBuiltin` refuse to retrieve the test runner, do we? (Side note, `module.builtinModules` probably _should_ include `node:test`, I’d think.)

I think the relevant equivalent is `module.isBuiltin`. Every input that returns `true` from that should be accessible from `process.getBuiltin`. I think most users would consider it a bug for `isBuiltin('node:test')` to return true and `getBuiltin('node:test')` to throw.

joyeecheung replied:
I think the more important question is, should `process.getBuiltin('test')` throw?

GeoffreyBooth replied:
> I think the more important question is, should `process.getBuiltin('test')` (and any future prefix-ed builtins) throw?

Yes, because `require('node:module').isBuiltin('test')` returns `false`. As in, there is no builtin named `test`. There’s only a builtin named `node:test`. It’s that simple, and it’s not a bug.

If you want to _additionally_ support `'test'` here, then it should get supported by `isBuiltin` first. The two APIs should support the same set of strings (as should `module.builtinModules`, but that’s a separate PR).

aduh95 replied:
Just my two cents, given we decided to always use the `node:` prefix for future built-in modules, wouldn't it be preferable to always require it? I.e., `process.getBuiltin('fs')` would throw, but `process.getBuiltin('node:fs')` would work – and `process.getBuiltin('node:test') would also work.

joyeecheung replied:
TBH I don't really know which to choose. I could go with no `node:`, always `node:`, or sometimes `node:`, but I am not super confident that they are the choices we won't regret about.....should we do a poll on this somehow?

GeoffreyBooth replied:
Why is this complicated? `isBuiltin` and `getBuiltin` should support exactly the same set of strings—no more, no less. Anything else would be inconsistent. And separately, we should fix the inconsistency of `builtinModules`. We don’t need a poll to know that our APIs should be consistent.

jasnell replied:
It's entirely likely that would implement support for this in worker's node.js compat mode. In that case, I would actually prefer the `node:` be allowed since we generally prefer the specificity.

JakobJingleheimer replied:
Consistency +++1

That said, `getBuiltin('node:')` is basically the same as `getBuiltin('builtin:')`, which seems a bit silly.

GeoffreyBooth replied:
It’s obviously redundant to write `node:` for `getBuiltin`, and I gather that that’s what troubles Joyee. But it’s quite a mouthful to document why `node:` is forbidden for `getBuiltin` and yet sometimes required for `isBuiltin`: 

> For `getBuiltin`, the `node:` prefix is unnecessary and is therefore forbidden. However for `isBuiltin`, the `node:` prefix is required for builtin modules that are only accessible with the prefix, such as `node:test`, to disambiguate from user modules. Therefore, `isBuiltin('node:test')` is true, but `getBuiltin('node:test')` throws; and `isBuiltin('test')` returns false while `getBuiltin('test')` returns the builtin.

This makes some sort of sense, kinda, but overall the end result feels like a bug, doesn’t it? Maybe if `getBuiltin` existed in isolation, it would feel natural to never need the prefix; but since it’s _not_ in isolation, we need to consider its counterparts. It’s much simpler to explain that there’s one set of strings of builtin IDs, and it’s the same set used by all three of `isBuiltin`, `getBuiltin` and `builtinModules`. In other words, this test should pass:

```js
const { ok } = require('assert');
const { builtinModules, isBuiltin, getBuiltin } = require('module');

const builtinIds = ['fs', 'node:fs', 'node:test'] // Plus everything where `isBuiltin(id) === true`
for (const id of builtinIds) {
  ok(builtinModules.includes(id));
  ok(isBuiltin(id));
  ok(getBuiltin(id));
  ok(process.getBuiltin(id));
}
```

There’s already an [open issue](https://github.com/nodejs/node/issues/42785) for `builtinModules` missing `node:test`, and from what I gather reading that issue there’s no debate about it; someone just needs to open a PR to add it (and also `node:`-prefixed versions of the older builtins like `fs`).

joyeecheung replied:
I updated it to accept the same kind of ID accepted by `module.isBuiltin()`. Though still looks weird to me, I don't think I really care about it that much as long as it's other people making that decision..

VoltrexKeyva commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+
+`process.getBuiltin(id)` provides a way to load the built-in modules
+in a globally available function. ES Modules that need to support
+other environments can use it to conditionally load a Node.js builtin
```
```suggestion
other environments can use it to conditionally load a Node.js built-in
```
For consistency.

VoltrexKeyva commented on lib/internal/modules/helpers.js:
```diff
@@ -342,8 +342,22 @@ let _hasStartedUserCJSExecution = false;
 let _hasStartedUserESMExecution = false;
 
+/**
+ * Load a public builtin.
+ * @param {string} id id of the builtin to be loaded. ID must be without the node: prefix.
+ * @returns {object} exports of the builtin.
```
```suggestion
 * Load a public built-in.
 * @param {string} id id of the built-in to be loaded. ID must be without the node: prefix.
 * @returns {object} exports of the built-in.
```
Ditto.

JakobJingleheimer commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+* `id` {string} ID of the built-in module being requested. Must not contain the `node:` prefix.
+* Returns: {Object}
+
+`process.getBuiltin(id)` provides a way to load the built-in modules
```
```suggestion
`process.getBuiltin(id)` provides a way to load built-in modules
```

JakobJingleheimer commented on doc/api/process.md:
````diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+asynchronous one.
+
+```mjs
+if (globalThis.process && globalThis.process.getBuiltin) {
````
nit

```suggestion
if (globalThis.process?.getBuiltin) {
```

JakobJingleheimer commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+`process.getBuiltin()` method returns the corresponding built-in
+module, which is the same as the object returned by
+[`require(id)`][`require()`]. If `id` does not correspond to any
+built-in module, an [`ERR_UNKNOWN_BUILTIN_MODULE`][] would be thrown.
```
```suggestion
built-in module, an [`ERR_UNKNOWN_BUILTIN_MODULE`][] is thrown.
```

jasnell replied:
I would actually prefer that the method not throw and return `undefined` in this case. I'm imagining a use case where another runtime can implement this method and it can be used to determine if that runtime actually provides an implementation of that built-in. For instance, in workers, we provide `node:stream` but not `node:fs` ... in the latter case, `process.getBuiltin(...)` returning `undefined` is a bit more optimal than creating an error that I'm likely just going to throw away and handle with a fallback.

joyeecheung replied:
Returning a falsy value SGTM, although I think the last time I looked into it I recall seeing an unrelated API returning null (unfortunately I cannot remember what API this is now...). I wonder if null would be better than undefined as that is a more consistent typeof.

jakebailey replied:
I'd personally prefer undefined over null... I'm not sure what you mean by "a more consistent typeof" when `typeof null === "object"`.

jakebailey replied:
That being said, I'm not sure why thowing is a big deal when that's already what would happen if you fail a require today...

joyeecheung replied:
undefined would be what my instincts say too but consistency seems important (but then, I cannot remember what the precedent is, maybe it was only something I saw in my dream somehow?)

joyeecheung replied:
Updated to return undefined for unfound modules.

JakobJingleheimer commented on lib/internal/modules/helpers.js:
```diff
@@ -342,8 +342,22 @@ let _hasStartedUserCJSExecution = false;
 
+/**
+ * Load a public builtin.
+ * @param {string} id id of the builtin to be loaded. ID must be without the node: prefix.
```
```suggestion
 * @param {string} id id of the builtin to be loaded. ID must be without the 'node:' prefix.
```

JakobJingleheimer reviewed (commented):
wow, quite simple :)

GeoffreyBooth commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+  // Run in Node.js, use the Node.js fs module.
+  const fs = globalThis.process.getBuiltin('fs');
+  // If `require()` is needed to load user-modules, use createRequire()
+  const m = globalThis.process.getBuiltin('module');
+  const require = m.createRequire(import.meta.url);
```
```suggestion
  const module = globalThis.process.getBuiltin('module');
  const require = module.createRequire(import.meta.url);
```

mcollina reviewed (approved):
lgtm

jakebailey commented:
This is obviously exactly what I asked for (😄) and gets it "done" the quickest, but I'm curious if there are any other opinions about this in general.

Is it possible that `import.now` or "weak/optional" imports come to fruition, such that those are more usable and processed by downstream tools? I'd hate to accidentally open the door to "breaking" node polyfills (node-polyfill-webpack-plugin, vite-plugin-node-polyfills, etc) and so on, just because it's not require or import, but some other expression.

joyeecheung commented:
>  I'd hate to accidentally open the door to "breaking" node polyfills (node-polyfill-webpack-plugin, vite-plugin-node-polyfills

Not quite sure if I am following this but I think `process.getBuiltin(id)` should be polyfill-able by these kind of bundler plugins - just patch the its polyfilled process object to load its own polyfilled libraries by id? (In general plugins like this already polyfill the process object).

jakebailey commented:
Yeah, I guess it's no different than detecting calls to `require` and error/warn on ones which cannot be statically determined. It's just not quite as injectable as `require`.

Just poking holes in my own idea.

joyeecheung commented:
We can also suggest in the documentation that users should only consider using this for code paths that must actually be run in Node.js (the code example is basically already suggesting that but we can make it clearer). If they want the code depending on the built-ins to be polyfill-able in the browser, they should still use require or import.

nicolo-ribaudo commented:
There is a TC39 proposal to get original built-in functions, through a new `getIntrinsics` global (https://github.com/tc39/proposal-get-intrinsic). `getBuiltns` is similar enough that _could_ cause confusion — what do you think about something like `process.getBuiltinModule` to be more clear about what it does?

MoLow commented on lib/internal/modules/helpers.js:
```diff
@@ -342,8 +342,22 @@ let _hasStartedUserCJSExecution = false;
+  if (!BuiltinModule.canBeRequiredByUsers(id)) {
+    throw new ERR_UNKNOWN_BUILTIN_MODULE(id);
+  }
+  return require(id);
```
```suggestion
  return require(`node:${id}`);
```
this seems more safe to me

MoLow replied:
unless we support passing ids that include the `node:` prefix

aduh95 replied:
Because internal modules are CJS modules, the `require` function that is available to them is not the same as the one available in userland CJS modules – one difference is that it doesn’t support `node:` prefix, another one is that it can only load built-in modules (including the ones in `lib/internal/`).

joyeecheung replied:
This is the internal require, it doesn’t recognize the node prefix.

jakebailey commented:
Just to prove this out to its full extent (and to procrastinate other work I didn't want to do), I polyfilled this API using `--require` and was able to get TypeScript's CLI, API, and test suite all working: https://github.com/microsoft/TypeScript/pull/58419

aduh95 commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,44 @@ console.log('After:', getActiveResourcesInfo());
+
+If `id` specifies a built-in available in the current Node.js process,
+`process.getBuiltin()` method returns the corresponding built-in
+module, which is the same as the object returned by
+[`require(id)`][`require()`]. If `id` does not correspond to any
```
Wouldn't it be more accurate to say it's the same as ``require(`node:${id}`)`` (i.e. it bypasses the `require` cache)
```suggestion
module, which is the same as the object returned by
[``require(`node:${id}`)``][`require()`]. If `id` does not correspond to any
```

https://github.com/nodejs/node/blob/4d59a9deda7c624d3341e3ee37452aee960f8816/doc/api/modules.md#L402-L405

joyeecheung commented:
Updated the PR a bit:

- Changed it to `process.getBuiltinModule()` as requested in https://github.com/nodejs/node/pull/52762#issuecomment-2088093800
- Return undefined for IDs without corresponding modules as requested in https://github.com/nodejs/node/pull/52762#discussion_r1590009551
- Accept IDs that are accepted by `module.isBuiltin` and reject otherwise (so this now throws on `getBuiltinModule('test')` etc) as requested in https://github.com/nodejs/node/pull/52762#discussion_r1585606673

GeoffreyBooth commented:
> Changed it to process.getBuiltinModule() as requested in process: add process.getBuiltinModule(id)

Now it mismatches `isBuiltin`. Personally I don't see the confusion between `getIntrinsic` and `getBuiltin`; I think `getBuiltin` is fine, and there's more potential for confusion with this not corresponding with `isBuiltin`.

Alternatively we can add `isBuiltinModule` as an alias for `isBuiltin`.

legendecas commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,45 @@ console.log('After:', getActiveResourcesInfo());
+by [`module.isBuiltin(id)`][]. Some built-in modules must be loaded with the
+`node:` prefix, see [built-in modules with mandatory `node:` prefix][].
+The built-in modules returned by `process.getBuiltinModule(id)` are always the
+original modules - that is, it's not affected by [`require.cache`][].
```
What does the statement "not affected by `require.cache`" suggest? AFAICT, the returned module can still be modified and may not be in a pristine state.

jakebailey replied:
Someone correct me if I'm wrong, but I believe it's possible for one to put a module directly into `require.cache`. So, if you were to set `fs` on it, `process.getBuiltinModule("fs")` will still get you the real one, while `require("fs")` would not. No matter what, patching the builtins still works, though, it's just the resolution from `require` that's affected by `require.cache`.

joyeecheung replied:
Modified the doc to say:

```md
The references returned by `process.getBuiltinModule(id)` always point to
the built-in module corresponding to `id` even if users modify
[`require.cache`][] in a way for `require(id)` to return something else.
```

Which should be clear enough. In any case it references the documentation of `require.cache` which talks about what that means in detail.

legendecas commented:
> Now it mismatches `isBuiltin`. Personally I don't see the confusion between `getIntrinsic` and `getBuiltin`; I think `getBuiltin` is fine, and there's more potential for confusion with this not corresponding with `isBuiltin`.
> 
> Alternatively we can add `isBuiltinModule` as an alias for `isBuiltin`.

`isBuiltin` is scoped in `node:module`. It is not a function available from the global scope, like `global.process.getBuiltinModule`.

GeoffreyBooth commented:
> `isBuiltin` is scoped in `node:module`. It is not a function available from the global scope, like `global.process.getBuiltinModule`.

Fair enough. I still think we should create an alias on `node:module` `isBuiltinModule`, but that can come later.

legendecas reviewed (approved):
LGTM. Just one small concern about docs that we should not make an impression that the result of `getBuiltinModule` can not be modified.

RedYetiDev reviewed (changes requested):
I'm not a core collaborator, so this is only a suggestion: Shouldn't this become a no-op with `--experimental-permission`?

joyeecheung commented:
> I'm not a core collaborator, so this is only a suggestion: Shouldn't this become a no-op with --experimental-permission?

I don't think it's relevant for the permission model as that guards things at the native layer when resources are actually accessed. The removed policy feature did guard access to builtins at the module loader level, but that's..removed. But cc @nodejs/security-wg to be sure.

RedYetiDev commented:
> > I'm not a core collaborator, so this is only a suggestion: Shouldn't this become a no-op with --experimental-permission?
> 
> I don't think it's relevant for the permission model as that guards things at the native layer when resources are actually accessed. The removed policy feature did guard access to builtins at the module loader level, but that's..removed (in general guarding things at the module loader level isn't super useful anyway because how widely monkey-patching of the module loader is used). But cc @nodejs/security-wg to be sure.

Thanks, I was thinking this was like `process.binding`, my bad, sorry

targos commented on doc/api/process.md:
````diff
@@ -1917,6 +1917,46 @@ console.log('After:', getActiveResourcesInfo());
+an asynchronous module, or turns a synchronous API into an asynchronous one.
+
+```mjs
+if (globalThis.process.getBuiltinModule) {
````
```suggestion
if (globalThis.process?.getBuiltinModule) {
```

nit: `process` might not be present in other environments.

targos commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,46 @@ console.log('After:', getActiveResourcesInfo());
+module. If `id` does not correspond to any built-in module, `undefined`
+is returned.
+
+`process.getBuiltinModule(id)` accept built-in module IDs that are recognized
```
```suggestion
`process.getBuiltinModule(id)` accepts built-in module IDs that are recognized
```

RedYetiDev reviewed (approved):
One nitpick, but other than that I like the idea.

> [!TIP]
> I'm a member of the triage team (and not a core collaborator), and my review, while it shows my support, has no power.

joyeecheung commented on doc/api/process.md:
```diff
@@ -1917,6 +1917,46 @@ console.log('After:', getActiveResourcesInfo());
+`node:` prefix, see [built-in modules with mandatory `node:` prefix][].
+The references returned by `process.getBuiltinModule(id)` always point to
+the built-in module corresponding to `id` even if users modify
+[`require.cache`][] in a way for `require(id)` to return something else.
```
I think that would be less clear as saying `if [require.cache][] has been modified to return something else.` implies that it's the returned value of `require.cache` being modified or `process.getBuiltinModule(id)` is modified, when it's actually `require()` that's affected, and the suggestion removed the mention of `require()` entirely.

RedYetiDev replied:
Whatever you think is best, its just "in a way for" feels off IMO.

nicolo-ribaudo replied:
"even if the user modifies `require.cache` so that `require(id)` returns something else"?

joyeecheung commented:
CI is green though this needs another approval from a collaborator to land @GeoffreyBooth @JakobJingleheimer @aduh95 @VoltrexKeyva @mcollina @Qard @jasnell @targos @MoLow @legendecas @Lxxyx @RafaelGSS @marco-ippolito

nicolo-ribaudo commented:
Would you be open to a PR backporting this to v20?

joyeecheung commented:
This has only been in the currently release for a week. We will need to bake it for another week before backporting it to LTS per the policy.
